## 前端错误捕获和处理

#### 1.异常分类（7种）

- Error：错误的基类，其他错误都继承此类型
- EvalError：Eval函数执行异常
- RangeError：数组越界 
- ReferenceError：尝试引用一个未被定义的变量时，会抛出此异常
- SyntaxError：语法解析不合理
- TypeError：类型错误，值的类型非预期类型
- URIError：以一种错误的方式使用全局URI处理函数抛出的异常

#### 2.异常处理和分析

- `try-catch`

  - `finally`子句是可选的，每次不管代码是否异常都会执行，`testFinally`方法永远返回“做作业”。try代码块中有代码错误时，catch中会捕获到。

    ```javascript
    function testFinally {
      try {
        return "出去玩";
      } catch (error) {
        return "看电视";
      } finally {
        return "做作业";
      }
      return "睡觉";
    }
    ```

    **但是`try-catch` 不能捕获语法错误和异步错误。**

    ```javascript
    try {
      setTimeout(() => {
        undefined.map(v => v);
      }, 1000)
    } catch(e) {
      console.log("捕获到异常：", e);
    }
     
    Uncaught TypeError: Cannot read property 'map' of undefined
      at <anonymous>:3:15
    ```

#### 3.异常捕获

##### window.error	

- Js运行时发生异常可捕获。但无法捕获静态资源加载和JS代码错误。

```javascript
window.onerror = function (message, source, lineno, colno, error) {
  console.log("捕获到异常：", { message, source, lineno, colno, error });
};
```

##### 静态资源加载异常

- onerror

  ```html
  <script>
    function errorHandler(error) {
      console.log("捕获到静态资源加载异常", error);
    }
  </script>
  <script src="http://cdn.xxx.com/js/test.js" onerror="errorHandler(this)"></script>
  <link rel="stylesheet" href="http://cdn.xxx.com/styles/test.css" onerror="errorHandler(this)">
  ```

- addEventListener("error")  网络请求异常不会事件冒泡，须在捕获阶段将其捕捉到，也就是在请求资源前，但不能判断HTTP的状态码。

  ```html
  <!DOCTYPE html>
  <html lang="zh">
      <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>error</title>
        <script>
          window.addEventListener('error', (error) => {
            console.log('捕获到异常：', error);
          }, true)
        </script>
      </head>
      <body>
        <img src="https://itemcdn.zcycdn.com/15af41ec-e6cb-4478-8fad-1a47402f0f25.png">
      </body>
   </html>
  ```

##### Promise异常

- Promise 中的异常不能被 try-catch 和 window.onerror 捕获，需要监听 unhandledrejection 来捕获这部分错误

  ```javascript
  window.addEventListener("unhandledrejection", function (e) {
    e.preventDefault();
    console.log("捕获到 promise 错误了");
    console.log("错误的原因是", e.reason);
    console.log("Promise 对象是", e.promise);
    return true;
  });
  
  Promise.reject("promise error");
  new Promise((resolve, reject) => {
    reject("promise error");
  });
  new Promise((resolve) => {
    resolve();
  }).then(() => {
    throw "promise error";
  });
  ```

##### React异常

-  React 组件是声明性的，try-catch 不太可靠，通过组件实现错误边界来捕获子组件树的任何错误。

  ```react
  class ErrorBoundary extends React.Component {
    constructor(props) {
      super(props);
      this.state = { hasError: false };
    }
  
    componentDidCatch(error, info) {
      // 展示出错的UI
      this.setState({ hasError: true });
      // 将错误信息上报到日志服务器
      logErrorToMyService(error, info);
    }
  
    render() {
      if (this.state.hasError) {
        // 可以展示自定义的错误样式
        return <h1>Something went wrong.</h1>;
      }
      return this.props.children;
    }
  }
  ```

  不会捕获以下错误：

  - 事件处理器
  - 异步代码
  - 服务端渲染代码
  - ErrorBoundary区域内的错误

  ```react
  <ErrorBoundary>
    <MyWidget />
  </ErrorBoundary
  ```

  

##### Vue异常

```vue
Vue.config.errorHandler = (err, vm, info) => {
  console.error("通过vue errorHandler捕获的错误");
  console.error(err);
  console.error(vm);
  console.error(info);
};
```



##### 请求异常

- axios 的拦截器来做统一梳理，同理能统一处理的异常也可以在放在拦截器里处理。

```javascript
// Add a response interceptor
axios.interceptors.response.use(
  function (response) {
    // Any status codes that falls outside the range of 2xx cause this function to trigger
    // Do something with response error
  },
  function (error) {
    if (error.response.status === 401) {
      goLogin(); // 跳转登录页
    } else if (error.response.status === 502) {
      alert(error.response.data.message || "系统升级中，请稍后重试");
    }
    return Promise.reject(error.response);
  }
);
```

#### 4.总结

异常一共七大类，处理时需分清是致命错误还是非致命错误。

- 可疑区域增加 `try-catch`
- 全局监控 `JS` 异常 `window.onerror`
- 全局监控静态资源异常 `window.addEventListener`
- 捕获没有 `catch` 的 `Promise` 异常用 `unhandledrejection`
- `Vue errorHandler` 和 `React componentDidCatch`
- `Axios` 请求统一异常处理用拦截器 `interceptors`
- 使用日志监控服务收集用户错误信息